crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -1156,6 +1188,214 @@ impl<'src> CodeExampleRst<'src> {
+        }
+
+        static FENCE_START: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(
+                r"(?xm)
+                ^
+                (?:
+                    # In the backtick case, info strings (following the fence)
+                    # cannot contain backticks themselves, since it would
+                    # introduce ambiguity with parsing inline code. In other
+                    # words, if we didn't specifically exclude matching `
+                    # in the info string for backtick fences, then we might
+                    # erroneously consider something to be a code fence block
+                    # that is actually inline code.
+                    #
+                    # NOTE: The `ticklang` and `tildlang` capture groups are
+                    # currently unused, but there was some discussion about not
+                    # assuming unlabeled blocks were Python. At the time of
+                    # writing, we do assume unlabeled blocks are Python, but
+                    # one could inspect the `ticklang` and `tildlang` capture
+                    # groups to determine whether the block is labeled or not.
+                    (?<ticks>```+)(?:\s*(?<ticklang>(?i:python|py|python3|py3))[^`]*)?
+                    |
+                    (?<tilds>~~~+)(?:\s*(?<tildlang>(?i:python|py|python3|py3))\p{any}*)?
+                )
+                $
+                ",
+            )
+            .unwrap()
+        });
+        let caps = FENCE_START.captures(rest)?;
+        let (fence_kind, fence_len) = if let Some(ticks) = caps.name("ticks") {
+            (MarkdownFenceKind::Backtick, ticks.as_str().chars().count())
+        } else {
+            let tildes = caps
+                .name("tilds")
+                .expect("no ticks means it must be tildes");
+            (MarkdownFenceKind::Tilde, tildes.as_str().chars().count())
+        };
+        Some(CodeExampleMarkdown {
+            lines: vec![],
+            opening_fence_indent: indentation_length(opening_fence_indent),
+            fence_kind,
+            fence_len,
+        })
+    }
+
+    /// Attempts to add the given line from a docstring to the Markdown code
+    /// snippet being collected.
+    ///
+    /// In this case, ownership is only not returned when the end of the block
+    /// was found, or if the block was determined to be invalid. A formatting
+    /// action is then pushed onto the queue.
+    fn add_code_line(
+        mut self,
+        original: InputDocstringLine<'src>,
+        queue: &mut VecDeque<CodeExampleAddAction<'src>>,
+    ) -> Option<CodeExampleMarkdown<'src>> {
+        if self.is_end(original) {
+            queue.push_back(self.into_format_action());
+            queue.push_back(CodeExampleAddAction::Print { original });
+            return None;
+        }
+        if !original.line.trim_whitespace().is_empty()
+            && indentation_length(original.line) < self.opening_fence_indent
```

Comment:
What is this case? The code is under-indented compared to the opening of the code fence? I actually don't know what Markdown does in that case.

Reply:
Yeah it's a weird case. I believe it is technically valid Markdown. But when I allowed it, it led to an idempotency bug that seemed a little tricky to fix. Basically, it seemed like a bug that was probably okay to ship with, and we can prioritize fixing it based on user feedback.

Good call out though. I added this comment:

```rust
        // When a line in a Markdown fenced closed block is indented *less*
        // than the opening indent, we treat the entire block as invalid.
        //
        // I believe that code blocks of this form are actually valid Markdown
        // in some cases, but the interplay between it and our docstring
        // whitespace normalization leads to undesirable outcomes. For example,
        // if the line here is unindented out beyond the initial indent of the
        // docstring itself, then this causes the entire docstring to have
        // its indent normalized. And, at the time of writing, a subsequent
        // formatting run undoes this indentation, thus violating idempotency.
```